### PR TITLE
Enable error strings when in a Debug build

### PR DESCRIPTION
**Pull request by zoddicus**

### CODE REVIEW

**Comment by zoddicus** [comment]
This is a first pass as enabling error strings in debug builds, using the existing macros, since it is clear to be how to rewrite things to get into a state of debug builds having strings, but release builds don't.

I feel like I am missing an element wrt the 'method switch' option you prefer. I get the general idea of having different method implementations based on the build config values instead of using the wrapper macro,  but I am unclear on how this will avoid including the string literals in the final artifact, which are the source of the size increase.

Is the idea that if the methods don't use the string, the optimizer is expected to remove the literals and keep binary size down?

**Comment by johnkslang** [comment]
No macros, no error5, etc.  And, this only needs to be about errors, no need to undo forwardCompatible or warnings, just the call to `error*`.

I was thinking just like the original code, except for the methods being something like

```
#ifdef GLSLANG_WEB && !_DEBUG
    virtual void C_DECL   error(const TSourceLoc&, const char* szReason, const char* szToken,
                                const char* szExtraInfoFormat, ...) { addError(); }
    virtual void C_DECL    warn(const TSourceLoc&, const char* szReason, const char* szToken,
                                const char* szExtraInfoFormat, ...) { }
    virtual void C_DECL ppError(const TSourceLoc&, const char* szReason, const char* szToken,
                                const char* szExtraInfoFormat, ...) { addError(); }
    virtual void C_DECL  ppWarn(const TSourceLoc&, const char* szReason, const char* szToken,
                                const char* szExtraInfoFormat, ...) { }
#else
    virtual void C_DECL   error(const TSourceLoc&, const char* szReason, const char* szToken,
                                const char* szExtraInfoFormat, ...);
    virtual void C_DECL    warn(const TSourceLoc&, const char* szReason, const char* szToken,
                                const char* szExtraInfoFormat, ...);
    virtual void C_DECL ppError(const TSourceLoc&, const char* szReason, const char* szToken,
                                const char* szExtraInfoFormat, ...);
    virtual void C_DECL  ppWarn(const TSourceLoc&, const char* szReason, const char* szToken,
                                const char* szExtraInfoFormat, ...);
#endif
```

Where other bodies and/or inheritance might need to be cleaned up a bit.

**Comment by zoddicus** [comment]
I have quickly hacked together, to the best of my understanding, a version of the method switching design you are detailing. I left in the macros and just made them pass through, since there were multiple patches that touched the error call sites, so I couldn't just revert a single patch. For a finished version I would also replace DEBUG with GLSLANG_WEB_DEVEL.

For this version, the WASM binary is 6kB larger then the previous version. I believe this is because the string literals being passed into the error methods are still being included in the binary even though they are not being used within the function. The optimizer appears to be unable to remove them, possibly because we are building as a library.

Do you have any suggestions on how to change this implementation to avoid this size increase?

**Comment by johnkslang** [comment]
Did this properly take care if inheritance?  When the function get overridden they are still alive?

Are the virtual methods disabling the compiler's ability to see what is dead?

I could see a combination of this with macros, if necessary, but it gets messy when different scopes are used: the macro approach breaks down.

I'm pretty sure I saw a savings doing something similar.

But, for now 6K is okay, for solving the immediate problem.

Do the errors show up correctly with this current commit?

File: glslang/MachineIndependent/ParseContextBase.cpp [hunk 1]
```diff
@@ -67,7 +67,7 @@ void TParseContextBase::outputMessage(const TSourceLoc& loc, const char* szReaso
     }
 }
 
-#ifndef GLSLANG_WEB
```
Comment (johnkslang) [inline glslang/MachineIndependent/ParseContextBase.cpp]: Where is `DEBUG` defined?

Shouldn't it be more like "KEEP_ERRORS" or "DEVEL"?  We are not making a debug build.

**Comment by johnkslang** [comment]
Ah, also see interactions with functions called by error-message strings, like `GetStorageQualifierString`, which was no longer needed when no errors are given.

File: glslang/MachineIndependent/ParseContextBase.cpp (hunk 1 above)
Comment (dneto0) [reply glslang/MachineIndependent/ParseContextBase.cpp]: I agree this should use a separate explicit control, other than whether it's a debug build.
I'm not too fussy about the name of such a thing.  Could be GLSLANG_WEB_WITH_ERR_STRINGS for the with-error-strings case.

**Comment by dneto0** [comment]
Given the function signatures in current form, with inheritance, I don't see how the compiler can remove the strings or other arguments.

I could see a middle ground where in the compact variation we turn off the inheritance and the base class provides an inline body which ignores the arguments.   The compiler would have good enough visibility that it would plausibly remove the string literal arguments.

Evaluation of the arguments still occurs because they may have side effects. So, e.g. GetStorageQualifierString is still called.

**Comment by johnkslang** [comment]
It is totally solvable either with macros (already done once) but preferably without (Ryan's trying to discover how).  If need to go into the reasons it's better done without macros I will.  If we must use macros we must, but I'm pretty sure (from doing similar things) we don't need them.

But, if people need errors now, whatever works on this experimental branch seems acceptable.

> Evaluation of the arguments still occurs because they may have side effects. So, e.g. GetStorageQualifierString is still called.

But, it should also be turned off (I'm pretty sure it is), via an empty body, and not have side effects.

It is a good point though, to look for those, as the macro idea truly threw them away, while the empty body idea does not.

BTW, the error body should not be truly empty, it should call `addError()`.  The compiler is not to change acceptance/rejection, it's just not to report the errors.

**Comment by johnkslang** [comment]
FYI, if we are truly wanting a bit more savings and *don't care about accept/reject* using an empty body can make the error tests themselves dead.

**Pull request closed without merging**